File: src/flake_analysis/api/schemas/clustering.py
"""Clustering schemas per backend design §1.2 + §1.3."""
from __future__ import annotations
from typing import Literal
from pydantic import BaseModel, Field


class SeedGroup(BaseModel):
    """A single seed group authored by the user."""
    name: str
    domain_ids: list[int]


class ClusteringRefitParams(BaseModel):
    """Mirrors pipeline/clustering.py:53-65."""
    seed_groups: list[SeedGroup]
    feature_cols: list[str] = Field(default_factory=lambda: ["mean_r", "mean_g", "mean_b"])
    covariance_type: Literal["full", "tied", "diag", "spherical"] = "full"
    rgb_threshold: float = 0.50
    fit_scope: Literal["seeds", "all_selected"] = "seeds"
    max_mahalanobis: float = 3.0


class ApplyThresholdsParams(BaseModel):
    """Mirrors pipeline/clustering.py:183-188."""
    cluster_thresholds: dict[int, float]
    max_mahalanobis: float | None = None


class ClusteringSummary(BaseModel):
    """Result wrapper used inside SSE 'done' event for /run/clustering/refit."""
    output_dir: str
    n_clusters: int
    n_assigned: int
    n_unassigned: int
    wrapper_params_hash: str | None = None


class ApplyThresholdsSummary(BaseModel):
    """Result wrapper for /run/clustering/apply_thresholds — mirrors apply_thresholds() return."""
    n_pass: int
    n_total: int
    n_clusters: int


class LabelsGroup(BaseModel):
    """One row of ``labels.json["groups"]`` per core/pipeline/clustering.py:272-286."""
    id: int
    name: str
    size: int
    mean_rgb: list[float]


class LabelsJson(BaseModel):
    """Frozen schema per core/pipeline/clustering.py:300-309 (plan v1 r7 §7.1)."""
    version: int
    n_clusters: int
    groups: list[LabelsGroup]
    assignments: dict[str, int]
    thresholds: dict[str, float]
    noise_label: int = -1
    random_state: int = 42
    fitted_at: str
    max_mahalanobis: float | None = None

File: web/src/api/clustering.ts
// web/src/api/clustering.ts
import { ApiError } from '@/api/selector'

async function unwrap<T>(resp: Response): Promise<T> {
  if (!resp.ok) {
    let envelope: any = null
    try {
      envelope = await resp.json()
    } catch {
      throw new ApiError(resp.status, 'http_error', `HTTP ${resp.status}`, null)
    }
    const err = envelope?.error ?? {}
    throw new ApiError(
      resp.status,
      err.code ?? 'http_error',
      err.message ?? `HTTP ${resp.status}`,
      err.details ?? null,
      err.request_id
    )
  }
  return (await resp.json()) as T
}

export interface SeedGroupDto {
  name: string
  domain_ids: number[]
}

export interface LabelsGroup {
  id: number
  name: string
  size: number
  mean_rgb: [number, number, number]
}

export interface LabelsJson {
  version: number
  n_clusters: number
  groups: LabelsGroup[]
  assignments: Record<string, number>
  thresholds: Record<string, number>
  noise_label: number
  random_state: number
  fitted_at: string
  max_mahalanobis?: number
}

export interface AssignmentsRows {
  domain_id: number[]
  cluster_label: number[]
  max_posterior: number[]
  nearest_mahalanobis?: number[]
  threshold_pass?: boolean[]
}

export interface ClusteringRefitBody {
  seed_groups: SeedGroupDto[]
  feature_cols?: string[]
  covariance_type?: 'full' | 'tied' | 'diag' | 'spherical'
  rgb_threshold?: number
  fit_scope?: 'seeds' | 'all_selected'
  max_mahalanobis?: number
}

export interface ApplyThresholdsBody {
  cluster_thresholds: Record<number, number>
  max_mahalanobis?: number | null
}

export async function fetchClusteringLabels(projectId: string): Promise<LabelsJson> {
  const resp = await fetch(
    `/api/v1/projects/${projectId}/data/clustering/labels`,
    { headers: { Accept: 'application/json' } }
  )
  return unwrap<LabelsJson>(resp)
}

export async function fetchClusteringAssignments(
  projectId: string
): Promise<AssignmentsRows> {
  const resp = await fetch(
    `/api/v1/projects/${projectId}/data/clustering/assignments`,
    { headers: { Accept: 'application/json' } }
  )
  return unwrap<AssignmentsRows>(resp)
}

export async function fetchClusteringSeedGroups(
  projectId: string
): Promise<SeedGroupDto[]> {
  const resp = await fetch(
    `/api/v1/projects/${projectId}/data/clustering/seed_groups`,
    { headers: { Accept: 'application/json' } }
  )
  return unwrap<SeedGroupDto[]>(resp)
}
